ap/tasks/reduction.py

```diff
@@ -76,30 +76,38 @@ def run(self):
             nano_file = input_file.load(formatter="uproot")
 
         # iterate over chunks of events and diffs
+        input_paths = [nano_file]
+        input_paths.append(inputs["sel"]["res"].path)
+        input_paths.extend([inp.path for inp in inputs["calib"]])
+        if self.selector_func.produced_columns:
+            input_paths.append(inputs["sel"]["cols"].path)
         with ChunkedReader(
-            [nano_file] + [inp.path for inp in inputs["calib"]] + [inputs["sel"]["res"].path],
-            source_type=["coffea_root", "awkward_parquet", "awkward_parquet"],
-            read_options=[{"iteritems_options": {"filter_name": load_columns_nano}}, None, None],
+            input_paths,
+            source_type=["coffea_root"] + (len(input_paths) - 1) * ["awkward_parquet"],
+            read_options=[{"iteritems_options": {"filter_name": load_columns_nano}}] + (len(input_paths) - 1) * [None],
         ) as reader:
             msg = f"iterate through {reader.n_entries} events ..."
-            for (events, *diffs, sel), pos in self.iter_progress(reader, reader.n_chunks, msg=msg):
+            for (events, sel, *diffs), pos in self.iter_progress(reader, reader.n_chunks, msg=msg):
                 # here, we would simply apply the mask from the selection results
                 # to filter events and objects
 
-                # add the calibrated diffs and new columns from selection results
-                events = update_ak_array(events, *(diffs + [sel.columns]))
+                # add the calibrated diffs and potentially new columns
+                events = update_ak_array(events, *diffs)
 
                 # add aliases
                 events = add_ak_aliases(events, aliases, remove_src=True)
 
                 # apply the event mask
-                events = events[sel.event]
+                event_mask = sel.event if "event" in sel.fields else Ellipsis
+                events = events[event_mask]
 
                 # apply all object masks whose names are present
+                # TODO: this should not be done automagically based on names as there might be
+                #       multiple collections of the same object type
                 for name in sel.objects.fields:
                     if name in events.fields:
                         # apply the event mask to the object mask first
-                        object_mask = sel.objects[name][sel.event]
+                        object_mask = sel.objects[name][event_mask]
```

Comment:
This depends on wether the selection masks in `selection` can be stacked, right? But I guess this shouldn't break even if the selections were stacked before 🤔

Reply:
OK :+1: